python_runner/runner.py

```diff
@@ -60,40 +62,40 @@ def callback(self, event_type, **data):
     def output(self, output_type, text, **extra):
         return self.output_buffer.put(output_type, text, **extra)
 
-    def execute(self, code_obj, source_code, mode=None):  # noqa
+    def execute(self, code_obj, mode=None):  # noqa
         return eval(code_obj, self.console.locals)  # noqa
 
     @contextmanager
-    def _execute_context(self, source_code):
+    def _execute_context(self):
         with self.output_buffer.redirect_std_streams():
             try:
                 yield
             except BaseException as e:
-                self.output("traceback", **self.serialize_traceback(e, source_code))
+                self.output("traceback", **self.serialize_traceback(e))
         self.post_run()
 
-    def run(self, source_code, mode="exec"):
-        code_obj = self.pre_run(source_code, mode)
-        with self._execute_context(source_code):
+    def run(self, source_code, mode="exec", top_level_await=False, filename=None):
```

Comment:
I don't think it makes sense to allow top_level_await here since the result isn't awaited.

`filename=None` also seems like a problem, doesn't that immediately break?

Reply:
Oh right, top_level_await indeed makes no sense. With the filename, I kind of want to make it optional to pass it when running code, so the filename stays the same unless explicitly changed. A check in the constructor could then prevent any Nones from causing trouble. Do you think that is sufficient?

Reply:
The check would need to be here, not in the constructor.